README.md
# oracle-secret [![Build Status](https://travis-ci.org/FWeinb/oracle-secret.svg?branch=master)](https://travis-ci.org/FWeinb/oracle-secret) [![Coverage Status](https://coveralls.io/repos/FWeinb/oracle-secret/badge.svg?branch=master&service=github)](https://coveralls.io/github/FWeinb/oracle-secret?branch=master)

> Encode/Decode Legend of Zelda Oracle of Ages and Oracle of Seasons secrets.

## Install

```
$ npm install --save oracle-secret
```

## Usage

Load a `GameSecret`
```js
import { GameSecret } from 'oracle-secret';

const secret = GameSecret.load('H~2:@ ←2♦yq GB3●( 6♥?↑6');

console.log(secret.toJSON());
//{
//  Hero: 'Link',
//  GameID: 14129,
//  Game: 'Ages',
//  Child: 'Pip',
//  Animal: 'Dimitri',
//  Behavior: 'BouncyD',
//  IsLinkedGame: true,
//  IsHeroQuest: false
//}
```

Create a `GameSecret`
```js
import { GameSecret, SecretParser } from 'oracle-secret';

const secret = new GameSecret({
  Hero: 'Zelda',
  GameID: 4000,
  Game: 'Ages',
  Child: 'Pip\u0000\u0000',
  Animal: 'Moosh',
  Behavior: 'HyperA',
  IsLinkedGame: true,
  IsHeroQuest: true
});

console.log(SecretParser.encode(secret.toBytes()));
// 8♥3↓! -756- rS/8T 9↓Y2d
```

## Special Thanks
 * [Andrew Nagle](https://github.com/kabili207) for his [oracle-hack](https://github.com/kabili207/oracle-hack) 




CHANGELOG.md
### 1.1.0
* Remove unnecessary `\u0000` from `Hero` and `Child`.
* Improve `SecretParser`.

### 1.0.0
* Inital release.


test/GameSecret.test.js
import assert from 'assert';
import { GameSecret } from '../src';


describe('GameSecret', () => {
  const secretString = 'H~2:@ ←2♦yq GB3●( 6♥?↑6';
  const secretBytes = [
    4, 37, 51, 36, 63,
    61, 51, 10, 44, 39,
    3,  0, 52, 21, 48,
    55,  9, 45, 59, 55,
  ];

  const secretJSON = {
    Hero: 'Link',
    Game: 'Ages',
    GameID: 14129,
    Child: 'Pip',
    Animal: 'Dimitri',
    Behavior: 'BouncyD',
    IsLinkedGame: true,
    IsHeroQuest: false,
  }

  describe('#load', () => {
    it('should load secret from byte array', () => {
      const secret = GameSecret.load(secretBytes);
      assert.deepEqual(
        secret.toJSON(),
        secretJSON
      );
    });
    it('should load secret from string', () => {
      const secret = GameSecret.load(secretString);
      assert.deepEqual(
        secret.toJSON(),
        secretJSON
      );
    });
    it('should throw when the secret is not 20 bytes', () => {
      assert.throws(
        GameSecret.load.bind(GameSecret, []),
        Error
      )
    });
    it('should throw when checksum is broken', () => {
      const brokenSecret = [
        4, 37, 51, 36, 63,
        61, 51, 10, 44, 39,
        3,  0, 52, 21, 48,
        55,  9, 45, /*Broken*/ 59 - 1, 55,
      ];
      assert.throws(
        GameSecret.load.bind(GameSecret, brokenSecret),
        Error
      )
    });
    it('should throw when it is a wrong secret type', () => {
      const brokenSecret = [
        59, 9, 52, 60, 20,
        25, 56, 54, 55, 25,
        40, 16, 35, 57, 17,
        58, 60, 19, 51, 29,
      ];
      assert.throws(
        GameSecret.load.bind(GameSecret, brokenSecret),
        Error
      )
    });
  });

  describe('#toBytes', () => {
    it('should produce a valid secret', () => {
      const secret = new GameSecret(secretJSON);
      assert.deepEqual(
        secret.toBytes(),
        secretBytes
      )
    });
  })

});


test/RingSecret.test.js
import assert from 'assert';
import { RingSecret } from '../src';

describe('RingSecret', () => {
  const secretString = 'L~2:N @bB↑& hmRh=';
  const secretBytes = [
    6, 37, 51, 36, 13,
    63, 26,  0, 59, 47,
    30, 32, 15, 30, 49,
  ];

  const secretJSON = {
    'GameID': 14129,
    'Rings': [
      'Power Ring L-1',
      'Dbl. Edge Ring',
      'Protection Ring',
    ],
  };

  describe('#load', () => {
    it('should load secret from byte array', () => {
      const secret = RingSecret.load(secretBytes);
      assert.deepEqual(
        secret.toJSON(),
        secretJSON
      );
    });
    it('should load secret from string', () => {
      const secret = RingSecret.load(secretString);
      assert.deepEqual(
        secret.toJSON(),
        secretJSON
      );
    });
    it('should throw when the secret is not 15 bytes', () => {
      assert.throws(
        RingSecret.load.bind(RingSecret, []),
        Error
      )
    });
    it('should throw when checksum is broken', () => {
      const brokenSecret = [
        6, 37, 51, 36, 13,
        63, 26,  0, 59, 47,
        30, 32, 15, /*Broken*/ 30 - 1, 49,
      ];
      assert.throws(
        RingSecret.load.bind(RingSecret, brokenSecret),
        Error
      )
    });
    it('should throw when it is a wrong secret type', () => {
      const brokenSecret = [
        0, 37, 51, 36, 13,
        63, 26,  0, 59, 47,
        30, 32, 15, 30, 63,
      ];
      assert.throws(
        RingSecret.load.bind(RingSecret, brokenSecret),
        Error
      )
    });
  });


  describe('#toBytes', () => {
    it('should produce a valid secret', () => {
      const secret = new RingSecret(secretJSON);
      assert.deepEqual(
        secret.toBytes(),
        secretBytes
      )
    });
  })

});


test/SecretParser.test.js
import assert from 'assert';
import * as SecretParser from '../src/SecretParser';

describe('SecretParser', () => {
  const secretString = 'H~2:@ ←2♦yq GB3●( 6♥?↑6';
  const secretPlaceholderString = 'H~2:@ {left}2♦yq GB3{circle}( 6{heart}?↑6';
  const secretBytes = [
    4,  37, 51, 36, 63,
    61, 51, 10, 44, 39,
    3,   0, 52, 21, 48,
    55,  9, 45, 59, 55,
  ];

  describe('#decode', () => {
    it('should work with valid secret', () => {
      assert.deepEqual(
        SecretParser.decode(secretString),
        secretBytes
      );
    });
    it('should work with placeholder', () => {
      assert.deepEqual(
        SecretParser.decode(secretPlaceholderString),
        secretBytes
      );
    });
    it('should throw when there is an invalid symbol', () => {
      assert.throws(
        SecretParser.decode.bind(SecretParser, 'ä'),
        Error
      )
    });
    it('should throw when there is an invalid placeholder', () => {
      assert.throws(
        SecretParser.decode.bind(SecretParser, '{green}'),
        Error
      )
    });
  });


  describe('#encode', () => {
    it('should work with valid bytes', () => {
      assert.equal(
        SecretParser.encode(secretBytes),
        secretString
      );
    });
    it('should throw when there is an invalid byte', () => {
      assert.throws(
        SecretParser.encode.bind(SecretParser, [228]),
        Error
      );
    });
  });

});


src/Secret.js
import BitArray from './BitArray';

const Cipher = [
  21, 35, 46,  4, 13, 63, 26, 16,
  58, 47, 30, 32, 15, 62, 54, 55,
  9,  41, 59, 49,  2, 22, 61, 56,
  40, 19, 52, 50,  1, 11, 10, 53,
  14, 27, 18, 44, 33, 45, 37, 48,
  25, 42,  6, 57, 60, 23, 51, 24,
];

export function decodeBytes(data) {
  var cipherKey = (data[0] >> 3);
  var cipherPosition = (cipherKey * 4);

  var secret = new Array(data.length);
  for (var i = 0; i < data.length; i++) {
    secret[i] = (data[i] ^ Cipher[cipherPosition++]);
  }

  secret[0] = (secret[0] & 7 | (cipherKey << 3));
  return secret;
}

export function calculateCipher(GameID) {
  return ((GameID >> 8) + (GameID & 255)) & 7
}

export function calculateChecksum(data, ignore) {
  return data.reduce((l,r,i) => i !== ignore ? l + r : l, 0) & 0xF;
}

export function byteToBinaryArray(data) {
  let bitArray = new BitArray(data.length * 6);
  let b = 0;
  for (var i=0;i<data.length;i++) {
    bitArray.setBeNumber(i * 6, data[i], 6);
  }
  return bitArray;
}

export function binaryToByteArray(bitArray) {
  let length = bitArray.length / 6;
  let arr = new Array(length);
  for (var i = 0; i < length; i++) {
    arr[i] = bitArray.readBe(i * 6, 6);
  }
  return arr;
}


src/BitArray.js
export default class BitArray  {

  constructor(size) {
    this._arr = Array.from({length: size}, () => 0);
  }

  get length() {
    return this._arr.length;
  }

  read(offset, length) {
    let value = 0;
    for (var i=0;i<length;i++) {
      value += this._arr[offset + i] * (1 << i);
    }
    return value;
  }

  readBe(offset, length) {
    let value = 0;
    let weight = 1 << (length - 1);
    for (var i=0;i<length;i++) {
      value += this._arr[offset + i] * weight;
      weight = weight >> 1;
    }
    return value;
  }

  set(offset, value) {
    this._arr[offset] = value;
    return this;
  }

  setArray(offset, arr) {
    for (var i=0;i<arr.length;i++) {
      this._arr[offset + i] = arr[i];
    }
    return this;
  }

  get(offset) {
    return this._arr[offset]
  }

  setLeNumber(offset, value, length) {
    let msbWeight = length - 1;
    for (var i=0;i<length;i++) {
      this._arr[offset + i] = (value >> i) & 1;
    }
    return this;
  }

  setBeNumber(offset, value, length) {
    let msbWeight = length - 1;
    for (var i=0;i<length;i++) {
      this._arr[offset + i] = (value >> (msbWeight - i)) & 1;
    }
    return this;
  }
}


src/GameSecret.js
import isString from 'is-string';
import BitArray from './BitArray';
import * as Secret from './Secret';
import { decode as decodeSecret } from './SecretParser'
import { createDualMapping } from './Helper'
import { Games, Animals, ChildBehaviors } from './Constants';

const GameMap = createDualMapping(Games);
const AnimalMap = createDualMapping(Animals);
const ChildBehaviorMap = createDualMapping(ChildBehaviors);

export default class GameSecret {

  constructor({
    Hero,
    GameID,
    Game,
    Child,
    Animal,
    Behavior,
    IsLinkedGame,
    IsHeroQuest,
  }) {
    this.Hero = Hero;
    this.GameID = GameID;
    this.Game = Game;
    this.Child = Child;
    this.Animal = Animal;
    this.Behavior = Behavior;
    this.IsLinkedGame = IsLinkedGame;
    this.IsHeroQuest = IsHeroQuest;
  }

  static load(secret) {
    if (isString(secret)) {
      secret = decodeSecret(secret);
    }

    if (secret.length !== 20) throw new Error('Length must be 20.')

    let decodedBytes  = Secret.decodeBytes(secret);
    let decodedSecret = Secret.byteToBinaryArray(decodedBytes);
    let checksum      = Secret.calculateChecksum(decodedBytes, 19);

    if (decodedBytes[19] != checksum) {
      throw new Error(`Checksum (${decodedBytes[19]}) does not match expected value (${checksum})`);
    }

    if (decodedSecret.get(3) != 0 || decodedSecret.get(4) != 0) {
      throw new Error('The specified data is not a game code');
    }

    var GameID       = decodedSecret.read(5, 15);
    var Game         = GameMap.valueFor(decodedSecret.get(21)+'');
    var IsHeroQuest  = decodedSecret.get(20)  === 1;
    var IsLinkedGame = decodedSecret.get(105) === 1;

    // Read Hero name
    var Hero = [ 22, 38, 60, 77, 89 ].reduce((l, r) => {
      return l + String.fromCharCode(decodedSecret.read(r, 8));
    }, '').replace(/\u0000/g,'');

    // Read Child name
    var Child = [ 30, 46, 68, 97, 106 ].reduce((l, r) => {
      return l + String.fromCharCode(decodedSecret.read(r, 8));
    }, '').replace(/\u0000/g,'');

    var Animal   = AnimalMap.valueFor(decodedSecret.read(85, 3));
    var Behavior = ChildBehaviorMap.valueFor(decodedSecret.read(54, 4));

    return new GameSecret({
      Hero,
      GameID,
      Game,
      Child,
      Animal,
      Behavior,
      IsLinkedGame,
      IsHeroQuest,
    });
  }

  toBytes() {
    var bitArray = new BitArray( 120 );

    bitArray.setLeNumber(0, Secret.calculateCipher(this.GameID), 3);

    bitArray.set(3, 0); // game secret
    bitArray.set(4, 0); // game secret

    bitArray.setLeNumber(5, this.GameID, 15);
    bitArray.set(20, this.IsHeroQuest ? 1 : 0);
    bitArray.set(21, GameMap.keyFor(this.Game));

    // Write Hero Name
    [ 22, 38, 60, 77, 89 ].forEach((v, i) => {
      bitArray.setLeNumber(v, this.Hero.charCodeAt(i), 8);
    });

    // Write Child Name
    [ 30, 46, 68, 97, 106 ].forEach((v, i) => {
      bitArray.setLeNumber(v, this.Child.charCodeAt(i), 8);
    });


    bitArray.setLeNumber(54, ChildBehaviorMap.keyFor(this.Behavior), 4);
    bitArray.set(76, 1); // unknown 1
    bitArray.setLeNumber(85, AnimalMap.keyFor(this.Animal), 3);
    bitArray.set(88, 1); // unknown 2
    bitArray.set(105, this.IsLinkedGame ? 1 : 0);

    var unencodedBytes = Secret.binaryToByteArray(bitArray);
    unencodedBytes[19] = Secret.calculateChecksum(unencodedBytes);

    return Secret.decodeBytes(unencodedBytes);
  }

  toJSON() {
    return {
      Hero:         this.Hero,
      Game:         this.Game,
      GameID:       this.GameID,
      Child:        this.Child,
      Animal:       this.Animal,
      Behavior:     this.Behavior,
      IsLinkedGame: this.IsLinkedGame,
      IsHeroQuest:  this.IsHeroQuest,
    };
  }
}


src/RingSecret.js
import isString from 'is-string';
import BitArray from './BitArray';
import * as Secret from './Secret';
import { decode as decodeSecret } from './SecretParser'
import { createDualMapping } from './Helper'
import { Rings } from './Constants';

const RingsMap = createDualMapping(Rings);

export default class RingSecret {

  constructor({GameID, Rings}) {
    this.GameID = GameID;
    this.Rings = Rings;
  }

  static load(secret) {

    if (isString(secret)) {
      secret = decodeSecret(secret);
    }

    if (secret.length !== 15) {
      throw new Error('Length must be 15.')
    }

    let decodedBytes  = Secret.decodeBytes(secret);
    let decodedSecret = Secret.byteToBinaryArray(decodedBytes);
    let checksum      = Secret.calculateChecksum(decodedBytes, 14);


    if (decodedBytes[14] != checksum) {
      throw new Error(`Checksum (${decodedBytes[14]}) does not match expected value (${checksum})`);
    }

    if (decodedSecret.get(3) != 0 || decodedSecret.get(4) != 1) {
      throw new Error('The specified data is not a ring code');
    }

    var GameID = decodedSecret.read(5, 15);

    var Rings = [
      52,53,54,55,56,57,58,59,
      20,21,22,23,24,25,26,27,
      68,69,70,71,72,73,74,75,
      44,45,46,47,48,49,50,41,
      60,61,62,63,64,65,66,67,
      28,29,30,31,32,33,34,35,
      76,77,78,79,80,81,82,83,
      36,37,38,39,40,41,42,43,
    ].map((v, i) => decodedSecret.get(v) === 1 ? RingsMap.valueFor(i) : null )
     .filter((v) => v !== null);

    return new RingSecret({
      GameID,
      Rings,
    });
  }

  toBytes() {
    var bitArray = new BitArray( 90 );


    var ringArr = Array.from({length: 64}, (v, k) => 0);

    this.Rings
    .map((value) => RingsMap.keyFor(value))
    .forEach((number) => {
      ringArr[number] = 1;
    });

    bitArray.setLeNumber(0, Secret.calculateCipher(this.GameID), 3);

    bitArray.set(3, 0); // game secret
    bitArray.set(4, 1); // game secret

    bitArray.setLeNumber(5, this.GameID, 15);

    bitArray.setArray(20, [
      8,   9, 10, 11, 12, 13, 14, 15,
      40, 41, 42, 43, 44, 45, 46, 47,
      56, 57, 58, 59, 60, 61, 62, 63,
      24, 25, 26, 27, 28, 29, 30, 31,
      0,   1,  2,  3,  4,  5,  6,  7,
      32, 33, 34, 35, 36, 37, 38, 39,
      16, 17, 18, 19, 20, 21, 22, 23,
      48, 49, 50, 51, 52, 53, 54, 55,
    ].map((v, i) => ringArr[v]));

    var unencodedBytes = Secret.binaryToByteArray(bitArray);
    unencodedBytes[14] = Secret.calculateChecksum(unencodedBytes);

    return Secret.decodeBytes(unencodedBytes);
  }

  toJSON() {
    return {
      GameID: this.GameID,
      Rings: this.Rings,
    }
  }

}


src/SecretParser.js
const Symbols = [
  'B', 'D', 'F', 'G', 'H', 'J', 'L', 'M', '♠', '♥', '♦', '♣', '#',
  'N', 'Q', 'R', 'S', 'T', 'W', 'Y', '!', '●', '▲', '■', '+', '-',
  'b', 'd', 'f', 'g', 'h', 'j',      'm', '$', '*', '/', ':', '~',
  'n', 'q', 'r', 's', 't', 'w', 'y', '?', '%', '&', '(', '=', ')',
  '2', '3', '4', '5', '6', '7', '8', '9', '↑', '↓', '←', '→', '@',
];

const nameSymbolMap = {
  'spade' : '♠',
  'heart': '♥',
  'diamond': '♦',
  'club': '♣',
  'circle': '●',
  'triangle': '▲',
  'square': '■',
  'up': '↑',
  'down': '↓',
  'left': '←',
  'right': '→',
}

const replacePattern = /\{(.*?)\}/gi

export function decode(secret) {
  secret = secret.replace(/\s+/g, '');
  secret = secret.replace(replacePattern, (fullMatch, match) => {
    let replace = nameSymbolMap[match];
    if (replace === undefined) throw new Error(`Placeholder ${match} not known.`);
    return replace;
  });
  var data = new Array(secret.length);

  for (var i = 0; i < secret.length; i++) {
    var symbol = Symbols.indexOf(secret[i]);
    if (symbol < 0 || symbol > 63)
      throw new Error('Secret contains invalid symbols');

    data[i] = symbol;
  }
  return data;
}

export function encode(data) {
  var str = '';
  for (var i = 0; i < data.length; i++) {
    if (data[i] < 0 || data[i] > 63) {
      throw new Error('Secret contains invalid values');
    }

    str += Symbols[data[i]];
    if (i % 5 == 4) {
      str += ' ';
    }
  }
  return str.trim();
}



src/Constants.js
const Rings = {
  0: 'Friendship Ring',
  1: 'Power Ring L-1',
  2: 'Power Ring L-2',
  3: 'Power Ring L-3',
  4: 'Armor Ring L-1',
  5: 'Armor Ring L-2',
  6: 'Armor Ring L-3',
  7: 'Red Ring',
  8: 'Blue Ring',
  9: 'Green Ring',
  10: 'Cursed Ring',
  11: 'Expert\'s Ring',
  12: 'Blast Ring',
  13: 'Rang Ring L-1',
  14: 'GBA Time Ring',
  15: 'Maple\'s Ring',
  16: 'Steadfast Ring',
  17: 'Pegasus Ring',
  18: 'Toss Ring',
  19: 'Heart Ring L-1',
  20: 'Heart Ring L-2',
  21: 'Swimmer\'s Ring',
  22: 'Charge Ring',
  23: 'Light Ring L-1',
  24: 'Light Ring L-2',
  25: 'Bomber\'s Ring',
  26: 'Green Luck Ring',
  27: 'Blue Luck Ring',
  28: 'Gold Luck Ring',
  29: 'Red Luck Ring',
  30: 'Green Holy Ring',
  31: 'Blue Holy Ring',
  32: 'Red Holy Ring',
  33: 'Snowshoe Ring',
  34: 'Roc\'s Ring',
  35: 'Quicksand Ring',
  36: 'Red Joy Ring',
  37: 'Blue Joy Ring',
  38: 'Gold Joy Ring',
  39: 'Green Joy Ring',
  40: 'Discovery Ring',
  41: 'Rang Ring L-2',
  42: 'Octo Ring',
  43: 'Moblin Ring',
  44: 'Like Like Ring',
  45: 'Subrosian Ring',
  46: 'First Gen Ring',
  47: 'Spin Ring',
  48: 'Bombproof Ring',
  49: 'Energy Ring',
  50: 'Dbl. Edge Ring',
  51: 'GBA Nature Ring',
  52: 'Slayer\'s Ring',
  53: 'Rupee Ring',
  54: 'Victory Ring',
  55: 'Sign Ring',
  56: '100th Ring',
  57: 'Whisp Ring',
  58: 'Gasha Ring',
  59: 'Peace Ring',
  60: 'Zora Ring',
  61: 'Fist Ring',
  62: 'Whimsical Ring',
  63: 'Protection Ring',
};

const Games = {
  0: 'Ages',
  1: 'Seasons',
};

const Animals = {
  3: 'Ricky',
  4: 'Dimitri',
  5: 'Moosh',
}

const ChildBehaviors = {
  0:  'Infant',
  1:  'BouncyA',
  2:  'BouncyB',
  3:  'BouncyC',
  4:  'BouncyD',
  5:  'BouncyE',
  6:  'ShyA',
  7:  'ShyB',
  8:  'ShyC',
  9:  'ShyD',
  10: 'ShyE',
  11: 'HyperA',
  12: 'HyperB',
  13: 'HyperC',
  14: 'HyperD',
  15: 'HyperE',
};



export {
  Rings,
  Games,
  Animals,
  ChildBehaviors,
};
